From fccc60abcd035b00f98d903066e4098c4ee1c294 Mon Sep 17 00:00:00 2001
From: Brenda Streiff <brenda.streiff@ni.com>
Date: Wed, 17 Nov 2021 12:57:06 -0600
Subject: [PATCH] opkg_conf: add pid suffix to volatile cache dir
To: opkg-devel@googlegroups.com

Volatile cache directories are supposed to have a lifetime scoped to a
particular process. Prior to feb40ec55c07 ("libopkg: do not require uid
0 for RO cmds"), opkg operations were protected by a single lock;
however, we might now have multiple opkg operations in-flight (for
example, "opkg print-architecture" running independently of an "opkg
install"). To handle this, uniquify the volatile cache_dir by appending
the process ID so that different opkg processes will not stomp over each
other.

Signed-off-by: Brenda Streiff <brenda.streiff@ni.com>
Upstream-Status: Accepted [https://groups.google.com/g/opkg-devel/c/x534f1NKOKk]
---
 libopkg/opkg_conf.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libopkg/opkg_conf.c b/libopkg/opkg_conf.c
index a8decef..37fcd2c 100644
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -854,7 +854,8 @@ int opkg_conf_load(void)
     }
 
     if (opkg_config->volatile_cache) {
-        sprintf_alloc(&tmp, "%s/%s", opkg_config->cache_dir, "volatile");
+        sprintf_alloc(&tmp, "%s/%s.%d", opkg_config->cache_dir, "volatile",
+                      (int)getpid());
         free(opkg_config->cache_dir);
         opkg_config->cache_dir = tmp;
     }
-- 
2.20.1

